File: src/features/typing/data/ts/n_queens.ts
function solveNQueens(n: number): string[][] {
    const result: string[][] = [];
    const queens: number[] = new Array(n).fill(0);
    const cols = new Set<number>();
    const diag1 = new Set<number>();
    const diag2 = new Set<number>();
    const backtrack = (row: number): void => {
        if (row === n) {
            const board: string[] = [];
            for (let i = 0; i < n; i++) {
                const line = new Array(n).fill('.');
                line[queens[i]] = 'Q';
                board.push(line.join(''));
            }
            result.push(board);
            return;
        }
        for (let col = 0; col < n; col++) {
            if (cols.has(col) || diag1.has(row + col) || diag2.has(row - col)) {
                continue;
            }
            queens[row] = col;
            cols.add(col);
            diag1.add(row + col);
            diag2.add(row - col);
            backtrack(row + 1);
            cols.delete(col);
            diag1.delete(row + col);
            diag2.delete(row - col);
        }
    };
    backtrack(0);
    return result;
}

File: src/features/typing/data/java/n_queens.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> result = new ArrayList<>();
        int[] queens = new int[n];
        Set<Integer> cols = new HashSet<>();
        Set<Integer> diag1 = new HashSet<>();
        Set<Integer> diag2 = new HashSet<>();
        backtrack(n, 0, queens, cols, diag1, diag2, result);
        return result;
    }

    private void backtrack(int n, int row, int[] queens, Set<Integer> cols, Set<Integer> diag1,
            Set<Integer> diag2, List<List<String>> result) {
        if (row == n) {
            List<String> board = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                StringBuilder line = new StringBuilder();
                for (int j = 0; j < n; j++) {
                    line.append('.');
                }
                line.setCharAt(queens[i], 'Q');
                board.add(line.toString());
            }
            result.add(board);
            return;
        }
        for (int col = 0; col < n; col++) {
            if (cols.contains(col) || diag1.contains(row + col) || diag2.contains(row - col)) {
                continue;
            }
            queens[row] = col;
            cols.add(col);
            diag1.add(row + col);
            diag2.add(row - col);
            backtrack(n, row + 1, queens, cols, diag1, diag2, result);
            cols.remove(col);
            diag1.remove(row + col);
            diag2.remove(row - col);
        }
    }
}
